timeout in TestAcceptQueueBacklogged test

```
=== RUN   TestAcceptQueueBacklogged
panic: test timed out after 10m0s

goroutine 890 [running]:
testing.(*M).startAlarm.func1()
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1700 +0x11f
created by time.goFunc
	/opt/hostedtoolcache/go/1.16.7/x64/src/time/sleep.go:180 +0x52

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000086600, 0xa5d6d9, 0x19, 0xb947b0, 0x1)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1239 +0x610
testing.runTests.func1(0xc000086600)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1511 +0xa7
testing.tRunner(0xc000086600, 0xc00013fce0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1193 +0x203
testing.runTests(0xc0000b6570, 0xe7e3c0, 0xa, 0xa, 0xc0464e872a20275b, 0x8bb31a1d67, 0xe85360, 0xc0000a2148)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1509 +0x613
testing.(*M).Run(0xc000140080, 0x0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1417 +0x3b4
main.main()
	_testmain.go:63 +0x237

goroutine 19 [select]:
github.com/ipfs/go-log/writer.(*MirrorWriter).logRoutine(0xc000097410)
	/home/runner/go/pkg/mod/github.com/ipfs/go-log@v1.0.4/writer/writer.go:71 +0x193
created by github.com/ipfs/go-log/writer.NewMirrorWriter
	/home/runner/go/pkg/mod/github.com/ipfs/go-log@v1.0.4/writer/writer.go:36 +0x13a

goroutine 13 [IO wait]:
internal/poll.runtime_pollWait(0x7f6a496d3880, 0x72, 0xe33610)
	/opt/hostedtoolcache/go/1.16.7/x64/src/runtime/netpoll.go:222 +0x55
internal/poll.(*pollDesc).wait(0xc000074198, 0x72, 0x1000, 0x1000, 0xffffffffffffffff)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_poll_runtime.go:87 +0xe6
internal/poll.(*pollDesc).waitRead(...)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc000074180, 0xc00007f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_unix.go:166 +0x28f
net.(*netFD).Read(0xc000074180, 0xc00007f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/net/fd_posix.go:55 +0x69
net.(*conn).Read(0xc000010028, 0xc00007f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/net/net.go:183 +0xec
bufio.(*Reader).fill(0xc000062600)
	/opt/hostedtoolcache/go/1.16.7/x64/src/bufio/bufio.go:101 +0x1af
bufio.(*Reader).ReadByte(0xc000062600, 0x38, 0x50, 0xc0001a2bf0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/bufio/bufio.go:253 +0x5d
github.com/multiformats/go-varint.ReadUvarint(0xc07020, 0xc000062600, 0xc000012320, 0xc000062780, 0xc0001a2c98)
	/home/runner/go/pkg/mod/github.com/multiformats/go-varint@v0.0.6/varint.go:80 +0xa2
github.com/libp2p/go-mplex.(*Multiplex).readNextHeader(0xc000074400, 0xc000012301, 0x0, 0x0, 0xc00007c080)
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:508 +0x65
github.com/libp2p/go-mplex.(*Multiplex).handleIncoming(0xc000074400)
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:365 +0x14b
created by github.com/libp2p/go-mplex.NewMultiplex
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:105 +0x3c5

goroutine 14 [select]:
github.com/libp2p/go-mplex.(*Multiplex).handleOutgoing(0xc000074400)
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:191 +0x145
created by github.com/libp2p/go-mplex.NewMultiplex
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:106 +0x3e7

goroutine 11 [IO wait]:
internal/poll.runtime_pollWait(0x7f6a496d3798, 0x72, 0xe33610)
	/opt/hostedtoolcache/go/1.16.7/x64/src/runtime/netpoll.go:222 +0x55
internal/poll.(*pollDesc).wait(0xc000074218, 0x72, 0x1000, 0x1000, 0xffffffffffffffff)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_poll_runtime.go:87 +0xe6
internal/poll.(*pollDesc).waitRead(...)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc000074200, 0xc00007e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_unix.go:166 +0x28f
net.(*netFD).Read(0xc000074200, 0xc00007e000, 0x1000, 0x1000, 0xc00007a400, 0x100, 0x0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/net/fd_posix.go:55 +0x69
net.(*conn).Read(0xc000010018, 0xc00007e000, 0x1000, 0x1000, 0x1, 0xc00007a400, 0x7f6a49e18308)
	/opt/hostedtoolcache/go/1.16.7/x64/src/net/net.go:183 +0xec
bufio.(*Reader).fill(0xc0000623c0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/bufio/bufio.go:101 +0x1af
bufio.(*Reader).ReadByte(0xc0000623c0, 0x0, 0x0, 0x4)
	/opt/hostedtoolcache/go/1.16.7/x64/src/bufio/bufio.go:253 +0x5d
github.com/multiformats/go-varint.ReadUvarint(0xc07020, 0xc0000623c0, 0x4, 0x4d231f, 0xc00004dc98)
	/home/runner/go/pkg/mod/github.com/multiformats/go-varint@v0.0.6/varint.go:80 +0xa2
github.com/libp2p/go-mplex.(*Multiplex).readNextHeader(0xc000074300, 0xc00004dd01, 0xc00004dd70, 0x1, 0x3)
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:508 +0x65
github.com/libp2p/go-mplex.(*Multiplex).handleIncoming(0xc000074300)
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:365 +0x14b
created by github.com/libp2p/go-mplex.NewMultiplex
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:105 +0x3c5

goroutine 12 [select]:
github.com/libp2p/go-mplex.(*Multiplex).handleOutgoing(0xc000074300)
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:191 +0x145
created by github.com/libp2p/go-mplex.NewMultiplex
	/home/runner/go/pkg/mod/github.com/libp2p/go-mplex@v0.3.0/multiplex.go:106 +0x3e7

goroutine 723 [chan receive]:
github.com/libp2p/go-libp2p-transport-upgrader.(*listener).Accept(0xc00043f2c0, 0x9d0820, 0xc00043f320, 0x10, 0x0)
	/home/runner/work/go-libp2p-transport-upgrader/go-libp2p-transport-upgrader/listener.go:149 +0x75
github.com/libp2p/go-libp2p-transport-upgrader_test.TestAcceptQueueBacklogged(0xc000435980)
	/home/runner/work/go-libp2p-transport-upgrader/go-libp2p-transport-upgrader/listener_test.go:307 +0x427
testing.tRunner(0xc000435980, 0xb947b0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1193 +0x203
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.16.7/x64/src/testing/testing.go:1238 +0x5d8

goroutine 724 [IO wait]:
internal/poll.runtime_pollWait(0x7f6a496d2cb8, 0x72, 0x72)
	/opt/hostedtoolcache/go/1.16.7/x64/src/runtime/netpoll.go:222 +0x55
internal/poll.(*pollDesc).wait(0xc00040ba98, 0x72, 0x0, 0x0, 0xa564c3)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_poll_runtime.go:87 +0xe6
internal/poll.(*pollDesc).waitRead(...)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Accept(0xc00040ba80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/internal/poll/fd_unix.go:401 +0x32f
net.(*netFD).accept(0xc00040ba80, 0x418710, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/net/fd_unix.go:172 +0x58
net.(*TCPListener).accept(0xc00027f350, 0xc0a120, 0xc00044c5d0, 0xe859e0)
	/opt/hostedtoolcache/go/1.16.7/x64/src/net/tcpsock_posix.go:139 +0x50
net.(*TCPListener).Accept(0xc00027f350, 0x43a6cf, 0xc0002aca87, 0x0, 0xc0001ccd50)
	/opt/hostedtoolcache/go/1.16.7/x64/src/net/tcpsock.go:261 +0x8e
github.com/multiformats/go-multiaddr/net.(*maListener).Accept(0xc0003462c0, 0x0, 0x0, 0xc00043f2c0, 0xc10970)
	/home/runner/go/pkg/mod/github.com/multiformats/go-multiaddr@v0.3.3/net/net.go:243 +0x71
github.com/libp2p/go-libp2p-transport-upgrader.(*listener).handleIncoming(0xc00043f2c0)
	/home/runner/work/go-libp2p-transport-upgrader/go-libp2p-transport-upgrader/listener.go:71 +0x158
created by github.com/libp2p/go-libp2p-transport-upgrader.(*Upgrader).UpgradeListener
	/home/runner/work/go-libp2p-transport-upgrader/go-libp2p-transport-upgrader/upgrader.go:48 +0x36b
FAIL	github.com/libp2p/go-libp2p-transport-upgrader	600.029s
FAIL

--- a/listener_test.go
+++ b/listener_test.go
@@ -6,6 +6,7 @@ import (
 	"net"
 	"os"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -285,33 +286,32 @@ func TestAcceptQueueBacklogged(t *testing.T) {
 	defer ln.Close()
 
 	// setup AcceptQueueLength connections, but don't accept any of them
-	errCh := make(chan error, st.AcceptQueueLength+1)
+	var counter int32 // to be used atomically
 	doDial := func() {
 		conn, err := dial(t, upgrader, ln.Multiaddr(), id)
-		errCh <- err
-		if conn != nil {
-			_ = conn.Close()
-		}
+		require.NoError(err)
+		atomic.AddInt32(&counter, 1)
+		t.Cleanup(func() { conn.Close() })
 	}
 
 	for i := 0; i < st.AcceptQueueLength; i++ {
 		go doDial()
 	}
 
-	require.Eventually(func() bool { return len(errCh) == st.AcceptQueueLength }, 2*time.Second, 100*time.Millisecond)
+	require.Eventually(func() bool { return int(atomic.LoadInt32(&counter)) == st.AcceptQueueLength }, 2*time.Second, 50*time.Millisecond)
 
 	// dial a new connection. This connection should not complete setup, since the queue is full
 	go doDial()
 
-	time.Sleep(500 * time.Millisecond)
-	require.Len(errCh, st.AcceptQueueLength)
+	time.Sleep(100 * time.Millisecond)
+	require.Equal(int(atomic.LoadInt32(&counter)), st.AcceptQueueLength)
 
 	// accept a single connection. Now the new connection should be set up, and fill the queue again
 	conn, err := ln.Accept()
 	require.NoError(err)
-	_ = conn.Close()
+	require.NoError(conn.Close())
 
-	require.Eventually(func() bool { return len(errCh) == st.AcceptQueueLength+1 }, 2*time.Second, 100*time.Millisecond)
+	require.Eventually(func() bool { return int(atomic.LoadInt32(&counter)) == st.AcceptQueueLength+1 }, 2*time.Second, 50*time.Millisecond)
 }
 
 func TestListenerConnectionGater(t *testing.T) {

